### DISCUSSION TITLE
Access nuxtApp from serverMiddleware

### DISCUSSION THREAD

**Discussion by FelixGrafschmidt** [nuxt/nuxt#17711] [Questions]
Hey all,

I am trying to authenticate users on the server since I will be using some secrets that I do not want to expose. These secrets are already available as e.g. nuxtApp.$config.secret.

What I am now trying to do is to access nuxtApp (or $config) in my serverMiddleware.

My attempt was this:

`import { useNuxtApp } from "#app"`

But this leads to this error:

``` 
WARN  Could not resolve import "#app" in C:\Users\Felix\Documents\dev\programming\mal-sort\server\middleware\token.ts using imports defined in C:\Users\Felix\Documents\dev\programming\mal-sort\package.json.       

WARN  '#app' is imported by server\middleware\token.ts, but could not be resolved – treating it as an external dependency
```

I already checked my tsconfig and I am correctly extending the one in ./.nuxt/tsconfig.json. In that file "#app" is listed as an alias for node_modules/nuxt3/dist/app, which is the correct path, afaik.

When I try to import from nuxt3 directly I receive the following error:

`import { useNuxtApp } from "nuxt3"`

```
 WARN  [worker] The requested module 'nuxt3' does not provide an export named 'useNuxtApp'       19:21:59  

 import { useNuxtApp } from 'nuxt3';
 ^^^^^^^^^^
 SyntaxError: The requested module 'nuxt3' does not provide an export named 'useNuxtApp'
 at ModuleJob._instantiate (node:internal/modules/esm/module_job:124:21)
 at async ModuleJob.run (node:internal/modules/esm/module_job:181:5)
 at async Promise.all (index 0)
 at async ESMLoader.import (node:internal/modules/esm/loader:281:24)
 at async loadESM (node:internal/process/esm_loader:88:5)
 at async handleMainPromise (node:internal/modules/run_main:65:12)
```

Both of the cases above are not marked in VSCode as being erroneuous in any way, the errors arise on build time.

Am I doing something wrong? I am using the latest nuxt3 prerelease with typescript.

**Comment by wobsoriano** [comment]
You'll have to access env variables using `process.env` in `middleware`. Though it is a known issue that it only works on development, and values are undefined on production.

See this https://github.com/nuxt/framework/discussions/1654

I think it's best if you create an issue from this discussion

**Reply by pi0** [reply]
This is not an issue process.env is not supposed to be preserved from build stage. In order to preserve config, sshould use runtimeConfig (import from `#config` or useRuntimeConfig)

**Reply by wobsoriano** [reply]
Edit: It works. I didn't know about `#config`.

```ts
// nuxt.config.ts
import { defineNuxtConfig } from 'nuxt3'

export default defineNuxtConfig({
    privateRuntimeConfig: {
        apiKey: process.env.API_KEY
    }
})
```

```ts
// server/middleware/test.ts
import runtimeConfig from '#config'

export default async (req, res) => {
  console.log(runtimeConfig. apiKey)
}
```

**Reply by reqwire** [reply]
I get Failed to resolve import "#config"

**Comment by AXSJ** [comment]
Hi, has this been answered at all? Config doesn't work for me. I need to access Pinia in server middleware so I need nuxtApp but I get the same error as stated in the original post

**Comment by reqwire** [comment]
Still no answer? I also need to access process.env variables from inside Pinia.
